### ISSUE TITLE
Add support for Kube-Prometheus-Stack monitoring

### ISSUE THREAD

**Issue by Luzifer** [ContinuousSecurityTooling/ots-helm-chart#10]
> As a systems operator I want to be able to monitor my applications.

With release `v1.10.0` OTS gained Prometheus metrics export which can be scraped by the Kube-Prometheus-Stack in an automated way without the OPS having to configure that themselves.

**Acceptance Criteria**
- A [ServiceMonitor](https://prometheus-operator.dev/docs/operator/api/#monitoring.coreos.com/v1.ServiceMonitor) is added to the chart scraping `/metrics` on the pods
- A Grafana Dashboard is built to visualize the [metrics exposed by OTS](https://github.com/Luzifer/ots/issues/142#issuecomment-1773947871) and [added as an annotated ConfigMap](https://fabianlee.org/2022/07/06/prometheus-adding-a-grafana-dashboard-using-a-configmap/)
- All of the above **must** be enabled through a flag in the `.Values` (default: `enabled = false`)
- When enabled the customization option `metricsAllowedSubnets` should automatically be enabled for cluster-local access (usual setup is to have the cluster run on `10.0.0.0/8` so allowing that should be sufficient though this almost certainly also exposes the metrics to the public so there needs to be a blocker in the ingress to block access to `/metrics`)
- A way to add [alerting rules](https://blog.r0b.io/post/getting-started-with-kube-prometheus-stack/#configuring-alertmanager) should be added but no default rules should be specified.

As a side-effect of enabling the `metricsAllowedSubnets` in customizations it might be a good idea to provide a map in the Values to set customization options which then are `{{ .Values.customizations | toYaml }}` rendered into a config-map (after being joined with the chart-set customizations) and mounted into the pod.

**Comment by hypery2k** [comment]
thanks for your suggestion. I wasn't aware that the metrics are now exposed.

Will try to incooperate the proposed changes next week

**Comment by hypery2k** [comment]
I made a first draft (Did not tried it yet). Also the grafana and alerting is missing.

Regarding the "subnet" allow list I'm a little unsure. I don't see a benefit of this setting. If you allow the service subnet it would also be accessible via the ingress. Then you have to also block it ingress-specific. Sounds a little to complicated.

Why not just checking the HTTP Request Header for `x-forwarded-host`? If the request gets through the ingress this value is set.

What do you think?

**Comment by Luzifer** [comment]
Well, not everything is Kubernetes and **never ever** trust that header without previously whitelisting the proxies which can send that header (which then would be just another setting configuring IPs / Subnets) aside of that in some scenarios you just don't have that header / the real IP. That setting is to - by default - cover up the metrics handler.

Just ensure that `/metrics` cannot be accessed through the Ingress and open the setting so a cluster-internal Prometheus can access it.

**Comment by hypery2k** [comment]
yeah, that's correct. The problem is that the there is no ingress-indepent configuration option for that

I made a simple, but not really elegent approach: https://github.com/ContinuousSecurityTooling/ots-helm-chart/blob/b098a83000e85e035f7b6430891c400d5fd4d822/chart/templates/ingress.yaml#L45-L52

**Comment by hypery2k** [comment]
@Luzifer Would it be okay for you to split the Monitoring + Grafana/Alerting features?

I'm currently not having that much time for the complete solution and would like to make to publish an update with this PR: https://github.com/ContinuousSecurityTooling/ots-helm-chart/pull/11

**Comment by hypery2k** [comment]
Metrics are scraped:
```
....
time="2025-01-21T11:04:13Z" level=info msg="10.42.22.92 - \"GET /metrics\" 200 2177 \"\" \"Prometheus/3.1.0\" 5.24456ms"
```

**Comment by Carapcoder** [comment]
Hello @hypery2k 

Please, could you share how you did to fix this issue with Prometheus? 
`expected a valid start token, got "\x1f" ("INVALID") while parsing: "\x1f"`

I'm having the same error when trying to scrape metrics from OTS.

**Comment by Luzifer** [comment]
@Carapcoder that should no longer happen with OTS v1.18.0 or later. In v1.18.0 the double-gzip-encoding bug was fixed.

**Comment by Carapcoder** [comment]
@Luzifer thank you very much!
I was using an older version than v1.18.x, when I deployed the latest version, it got fixed!
